docs/configuration.md
## Configuration for admin

<img alt="Tortoise" src="images/eating.jpg" width="400px"/>

The cluster admin can set the global configurations via the configuration file.
The configuration file is passed via `--config` flag.

```
RangeOfMinMaxReplicasRecommendationHours:     The time (hours) range of minReplicas and maxReplicas recommendation (default: 1)
GatheringDataPeriodType:                      How long do we gather data for minReplica/maxReplica or data from VPA. "daily" and "weekly" are only valid value. (default: weekly)
TTLHoursOfMinMaxReplicasRecommendation:       The TTL of minReplicas and maxReplicas recommendation (default: 720 (=30 days))
MaxReplicasFactor:                            The factor to calculate the maxReplicas recommendation from the current replica number (default: 2.0)
MinReplicasFactor:                            The factor to calculate the minReplicas recommendation from the current replica number (default: 0.5)
ReplicaReductionFactor:                       The factor to reduce the minReplicas gradually after turning off Emergency mode (default: 0.95)
UpperTargetResourceUtilization:               The max target utilization that tortoise can give to the HPA (default: 90)
MinimumMinReplicas:                           The minimum minReplicas that tortoise can give to the HPA (default: 3)
PreferredReplicaNumUpperLimit:                The replica number which the tortoise tries to keep the replica number less than. As it says "preferred", the tortoise **tries** to keep the replicas number less than this, but the replica number may be more than this when other "required" rule will be violated by this limit. (default: 30)
MaximumCPUCores:                              The maximum CPU cores that the tortoise can give to the container (default: 10)
MaximumMemoryBytes:                           The maximum memory bytes that the tortoise can give to the container (default: 10Gi)
Timezone:                                     The timezone used to record time in tortoise objects (default: Asia/Tokyo)
TortoiseUpdateInterval:                       The interval of updating each tortoise (default: 15s)
TortoiseHPATargetUtilizationMaxIncrease:      The max increase of target utilization that tortoise can give to the HPA (default: 5)
TortoiseHPATargetUtilizationUpdateInterval:   The interval of updating target utilization of each HPA (default: 1h)
```

### RangeOfMinMaxReplicasRecommendationHours

The time (hours) range of minReplicas and maxReplicas recommendation (default: 1)

```yaml
kind: Tortoise
#...
status: 
  recommendations:
    horizontal:
      minReplicas:
        - from: 0
          to: 1
          weekday: Sunday
          timezone: Asia/Tokyo
          value: 3
          updatedAt: 2023-01-01T00:00:00Z
        - from: 1
          to: 2
          weekday: Sunday
          timezone: Asia/Tokyo
          value: 3
          updatedAt: 2023-01-01T00:00:00Z
```

### GatheringDataPeriodType

GatheringDataPeriodType means how long do we gather data for minReplica/maxReplica or data from VPA. "daily" and "weekly" are only valid value. (default: weekly)
If "daily", tortoise will consider all workload behaves very similarly every day.
If your workload may behave differently on, for example, weekdays and weekends, set this to "weekly".

#### "daily"

```yaml
kind: Tortoise
#...
status: 
  recommendations:
    horizontal:
      minReplicas:
        # This recommendation is from 0am to 1am on all days of week.
        - from: 0
          to: 1
          timezone: Asia/Tokyo
          value: 3
          updatedAt: 2023-01-01T00:00:00Z
        - from: 1
          to: 2
          timezone: Asia/Tokyo
          value: 3
          updatedAt: 2023-01-01T00:00:00Z
        # ...
        - from: 23
          to: 24
          timezone: Asia/Tokyo
          value: 3
          updatedAt: 2023-01-01T00:00:00Z
```

#### "weekly"

```yaml
kind: Tortoise
#...
status: 
  recommendations:
    horizontal:
      minReplicas:
        # This recommendation is from 0am to 1am on Sundays.
        - from: 0
          to: 1
          weekday: Sunday # Recommendation is generated for each day of week.
          timezone: Asia/Tokyo
          value: 3
          updatedAt: 2023-01-01T00:00:00Z
        - from: 1
          to: 2
          weekday: Sunday
          timezone: Asia/Tokyo
          value: 3
          updatedAt: 2023-01-01T00:00:00Z
        # ...
        - from: 23
          to: 24
          weekday: Saturday
          timezone: Asia/Tokyo
          value: 3
          updatedAt: 2023-01-01T00:00:00Z
```

### TTLHoursOfMinMaxReplicasRecommendation

The TTL of minReplicas and maxReplicas recommendation (default: 720 (=30 days))

```yaml
kind: Tortoise
#...
status: 
  recommendations:
    horizontal:
      minReplicas:
        - from: 0
          to: 1
          weekday: Sunday
          timezone: Asia/Tokyo
          value: 3
          updatedAt: 2023-01-01T00:00:00Z  # this recommendation will be expired on 2023-01-31.
```

### MaxReplicasFactor

The factor to calculate the maxReplicas recommendation from the current replica number (default: 2.0)

If the current replica number is 15 and `MaxReplicasFactor` is 2.0,
the maxReplicas recommendation from the current situation will be 30 (15 * 2.0).

```yaml
kind: Tortoise
#...
status: 
  recommendations:
    horizontal:
      maxReplicas:
        - from: 0
          to: 1
          weekday: Sunday
          timezone: Asia/Tokyo
          value: 30 
          updatedAt: 2023-01-01T00:00:00Z  
```

### MinReplicasFactor

The factor to calculate the minReplicas recommendation from the current replica number (default: 0.5)

If the current replica number is 10 and `MaxReplicasFactor` is 0.5,
the minReplicas recommendation from the current situation will be 5 (10 * 0.5).

```yaml
kind: Tortoise
#...
status: 
  recommendations:
    horizontal:
      minReplicas:
        - from: 0
          to: 1
          weekday: Sunday
          timezone: Asia/Tokyo
          value: 5
          updatedAt: 2023-01-01T00:00:00Z  
```

### ReplicaReductionFactor

The factor to reduce the minReplicas gradually after turning off Emergency mode (default: 0.95)

Let's say `ReplicaReductionFactor` is 0.95,
the minReplicas was increased to 100 due to the emergency mode,
and a user just turned off the emergency mode now.

Then, the `minReplicas` is going to change like:

100 --(*0.95)--> 95 --(*0.95)--> 91 -- ...

It's reduced every time tortoise is evaluated by the controller. (= once a `TortoiseUpdateInterval`)

### UpperTargetResourceUtilization

The max target utilization that tortoise can give to the HPA (default: 90)
So, HPA target utilization managed by tortoise won't be higher than this value.

### MinimumMinReplicas

The minimum minReplicas that tortoise can give to the HPA (default: 3)
So, HPA minReplicas managed by tortoise won't be smaller than this value.

### PreferredReplicaNumUpperLimit

The replica number which the tortoise tries to keep the replica number less than. (default: 30)

As it says "preferred", the tortoise **tries** to keep the replicas number less than this, 
but the replica number may be more than this when other "required" rules (`MaximumCPUCores` and `MaximumMemoryBytes`) will be violated by this limit. 

So, when the number of replicas reaches `PreferredReplicaNumUpperLimit`,
a tortoise will increase the Pod's resource request instead of increasing the number of replicas.

But, when the resource request reaches `MaximumCPUCores` or `MaximumMemoryBytes`,
a tortoise will ignore `PreferredReplicaNumUpperLimit`, and increase the number of replicas.

### MaximumCPUCores

The maximum CPU cores that the tortoise can give to the container (default: 10)

Note that it's the upper limit for the container, not for the Pod.

### MaximumMemoryBytes

The maximum memory bytes that the tortoise can give to the container (default: 10Gi)

Note that it's the upper limit for the container, not for the Pod.

### timezone

The timezone which used to record time in tortoise objects (default: Asia/Tokyo)

### TortoiseUpdateInterval 

The interval of updating each tortoise (default: 15s)

But, it may delay if there are many tortoise objects in the cluster.

### TortoiseHPATargetUtilizationMaxIncrease

The max increase of target utilization that tortoise can give to the HPA (default: 5)

If tortoise suggests changing the HPA target resource utilization from 50 to 80, it might be dangerous to give the change at once.
By configuring this, we can limit the max increase that tortoise can make. 
So, if TortoiseHPATargetUtilizationMaxIncrease is 5, even if tortoise suggests changing the HPA target resource utilization from 50 to 80, 
the target utilization is actually change from 50 to 55.

### TortoiseHPATargetUtilizationUpdateInterval

The interval of updating target utilization of each HPA (default: 1h)

So, similarily to TortoiseHPATargetUtilizationMaxIncrease, it's also a safety guard to prevent HPA target utilization from suddenly changed.
If TortoiseHPATargetUtilizationMaxIncrease is 1h, TortoiseHPATargetUtilizationMaxIncrease is 5, and tortoise keep suggesting changing the HPA target resource utilization from 50 to 80,
the target resource utilization would be changing like 50 -(1h)-> 55 -(1h)-> 60 → ... → 80.

docs/flag-configuration.md
## Flag configurations for admin

<img alt="Tortoise" src="images/eating.jpg" width="400px"/>

The cluster admin can set the global configurations via flags.

```
--range-of-min-max-replicas-recommendation-hours: The time (hours) range of minReplicas and maxReplicas recommendation (default: 1)
--ttl-hours-of-min-max-replicas-recommendation:   The TTL of minReplicas and maxReplicas recommendation (default: 720 (=30 days))
--max-replicas-factor:                            The factor to calculate the maxReplicas recommendation from the current replica number (default: 2.0)
--min-replicas-factor:                            The factor to calculate the minReplicas recommendation from the current replica number (default: 0.5)
--replica-reduction-factor:                       The factor to reduce the minReplicas gradually after turning off Emergency mode (default: 0.95)
--upper-target-resource-utilization:              The max target utilization that tortoise can give to the HPA (default: 90)
--minimum-min-replicas:                           The minimum minReplicas that tortoise can give to the HPA (default: 3)
--preferred-replicas-number-upper-limit:          The replica number which the tortoise tries to keep the replica number less than. As it says "preferred", the tortoise **tries** to keep the replicas number less than this, but the replica number may be more than this when other "required" rule will be violated by this limit. (default: 30)
--maximum-cpu-cores:                              The maximum CPU cores that the tortoise can give to the container (default: 10)
--maximum-memory-bytes:                           The maximum memory bytes that the tortoise can give to the container (default: 10Gi)
--timezone:                                       The timezone used to record time in tortoise objects (default: Asia/Tokyo)
--tortoise-update-interval:                       The interval of updating each tortoise (default: 15s)
```

### range-of-min-max-replicas-recommendation-hours

The time (hours) range of minReplicas and maxReplicas recommendation (default: 1)

```yaml
kind: Tortoise
#...
status: 
  recommendations:
    horizontal:
      minReplicas:
        - from: 0
          to: 1
          weekday: Sunday
          timezone: Asia/Tokyo
          value: 3
          updatedAt: 2023-01-01T00:00:00Z
        - from: 1
          to: 2
          weekday: Sunday
          timezone: Asia/Tokyo
          value: 3
          updatedAt: 2023-01-01T00:00:00Z
```

### ttl-hours-of-min-max-replicas-recommendation

The TTL of minReplicas and maxReplicas recommendation (default: 720 (=30 days))

```yaml
kind: Tortoise
#...
status: 
  recommendations:
    horizontal:
      minReplicas:
        - from: 0
          to: 1
          weekday: Sunday
          timezone: Asia/Tokyo
          value: 3
          updatedAt: 2023-01-01T00:00:00Z  # this recommendation will be expired on 2023-01-31.
```

### max-replicas-factor

The factor to calculate the maxReplicas recommendation from the current replica number (default: 2.0)

If the current replica number is 15 and `max-replicas-factor` is 2.0,
the maxReplicas recommendation from the current situation will be 30 (15 * 2.0).

```yaml
kind: Tortoise
#...
status: 
  recommendations:
    horizontal:
      maxReplicas:
        - from: 0
          to: 1
          weekday: Sunday
          timezone: Asia/Tokyo
          value: 30 
          updatedAt: 2023-01-01T00:00:00Z  
```

### min-replicas-factor

The factor to calculate the minReplicas recommendation from the current replica number (default: 0.5)

If the current replica number is 10 and `max-replicas-factor` is 0.5,
the minReplicas recommendation from the current situation will be 5 (10 * 0.5).

```yaml
kind: Tortoise
#...
status: 
  recommendations:
    horizontal:
      minReplicas:
        - from: 0
          to: 1
          weekday: Sunday
          timezone: Asia/Tokyo
          value: 5
          updatedAt: 2023-01-01T00:00:00Z  
```

### replica-reduction-factor

The factor to reduce the minReplicas gradually after turning off Emergency mode (default: 0.95)

Let's say `replica-reduction-factor` is 0.95,
the minReplicas was increased to 100 due to the emergency mode,
and a user just turned off the emergency mode now.

Then, the `minReplicas` is going to change like:

100 --(*0.95)--> 95 --(*0.95)--> 91 -- ...

It's reduced every time tortoise is evaluated by the controller. (= once a `tortoise-update-interval`)

### upper-target-resource-utilization

The max target utilization that tortoise can give to the HPA (default: 90)
So, HPA target utilization managed by tortoise won't be higher than this value.

### minimum-min-replicas

The minimum minReplicas that tortoise can give to the HPA (default: 3)
So, HPA minReplicas managed by tortoise won't be smaller than this value.

### preferred-replicas-number-upper-limit

The replica number which the tortoise tries to keep the replica number less than. (default: 30)

As it says "preferred", the tortoise **tries** to keep the replicas number less than this, 
but the replica number may be more than this when other "required" rules (`maximum-cpu-cores` and `maximum-memory-bytes`) will be violated by this limit. 

So, when the number of replicas reaches `preferred-replicas-number-upper-limit`,
a tortoise will increase the Pod's resource request instead of increasing the number of replicas.

But, when the resource request reaches `maximum-cpu-cores` or `maximum-memory-bytes`,
a tortoise will ignore `preferred-replicas-number-upper-limit`, and increase the number of replicas.

### maximum-cpu-cores

The maximum CPU cores that the tortoise can give to the container (default: 10)

Note that it's the upper limit for the container, not for the Pod.

### maximum-memory-bytes

The maximum memory bytes that the tortoise can give to the container (default: 10Gi)

Note that it's the upper limit for the container, not for the Pod.

### timezone

The timezone which used to record time in tortoise objects (default: Asia/Tokyo)

### tortoise-update-interval 

The interval of updating each tortoise (default: 15s)

But, it may delay if there are many tortoise objects in the cluster.


pkg/config/config.go
package config

import (
	"fmt"
	"os"
	"time"

	"gopkg.in/yaml.v3"
)

type Config struct {
	// RangeOfMinMaxReplicasRecommendationHours is the time (hours) range of minReplicas and maxReplicas recommendation (default: 1)
	RangeOfMinMaxReplicasRecommendationHours int `yaml:"RangeOfMinMaxReplicasRecommendationHours"`
	// GatheringDataPeriodType means how long do we gather data for minReplica/maxReplica or data from VPA. "daily" and "weekly" are only valid value. (default: weekly)
	// If "daily", tortoise will consider all workload behaves very similarly every day.
	// If your workload may behave differently on, for example, weekdays and weekends, set this to "weekly".
	GatheringDataPeriodType string `yaml:"GatheringDataPeriodType"`
	// TTLHoursOfMinMaxReplicasRecommendation is the TTL (hours) of minReplicas and maxReplicas recommendation (default: 720 (=30 days))
	TTLHoursOfMinMaxReplicasRecommendation int `yaml:"TTLHoursOfMinMaxReplicasRecommendation"`
	// MaxReplicasFactor is the factor to calculate the maxReplicas recommendation from the current replica number (default: 2.0)
	MaxReplicasFactor float64 `yaml:"MaxReplicasFactor"`
	// MinReplicasFactor is the factor to calculate the minReplicas recommendation from the current replica number (default: 0.5)
	MinReplicasFactor float64 `yaml:"MinReplicasFactor"`
	// ReplicaReductionFactor is the factor to reduce the minReplicas gradually after turning off Emergency mode (default: 0.95)
	ReplicaReductionFactor float64 `yaml:"ReplicaReductionFactor"`
	// UpperTargetResourceUtilization is the max target utilization that tortoise can give to the HPA (default: 90)
	UpperTargetResourceUtilization int `yaml:"UpperTargetResourceUtilization"`
	// MinimumMinReplicas is the minimum minReplicas that tortoise can give to the HPA (default: 3)
	MinimumMinReplicas int `yaml:"MinimumMinReplicas"`
	// PreferredReplicaNumUpperLimit is the replica number which the tortoise tries to keep the replica number less than. As it says "preferred", the tortoise **tries** to keep the replicas number less than this, but the replica number may be more than this when other "required" rule will be violated by this limit. (default: 30)
	PreferredReplicaNumUpperLimit int `yaml:"PreferredReplicaNumUpperLimit"`
	// MaximumCPUCores is the maximum CPU cores that the tortoise can give to the container (default: 10)
	MaximumCPUCores string `yaml:"MaximumCPUCores"`
	// MaximumMemoryBytes is the maximum memory bytes that the tortoise can give to the container (default: 10Gi)
	MaximumMemoryBytes string `yaml:"MaximumMemoryBytes"`
	// TimeZone is the timezone used to record time in tortoise objects (default: Asia/Tokyo)
	TimeZone string `yaml:"TimeZone"`
	// TortoiseUpdateInterval is the interval of updating each tortoise (default: 15s)
	TortoiseUpdateInterval time.Duration `yaml:"TortoiseUpdateInterval"`
	// TortoiseHPATargetUtilizationMaxIncrease is the max increase of target utilization that tortoise can give to the HPA (default: 5)
	TortoiseHPATargetUtilizationMaxIncrease int `yaml:"TortoiseHPATargetUtilizationMaxIncrease"`
	// TortoiseHPATargetUtilizationUpdateInterval is the interval of updating target utilization of each HPA (default: 1h)
	TortoiseHPATargetUtilizationUpdateInterval time.Duration `yaml:"TortoiseHPATargetUtilizationUpdateInterval"`

	// TODO: the following fields should be removed after we stop depending on deployment.
	// So, we don't put them in the documentation.
	// IstioSidecarProxyDefaultCPU is the default CPU resource request of the istio sidecar proxy (default: 100m)
	IstioSidecarProxyDefaultCPU string `yaml:"IstioSidecarProxyDefaultCPU"`
	// IstioSidecarProxyDefaultMemory is the default Memory resource request of the istio sidecar proxy (default: 200Mi)
	IstioSidecarProxyDefaultMemory string `yaml:"IstioSidecarProxyDefaultMemory"`
}

// ParseConfig parses the config file (yaml) and returns Config.
func ParseConfig(path string) (*Config, error) {
	config := &Config{
		RangeOfMinMaxReplicasRecommendationHours:   1,
		GatheringDataPeriodType:                    "weekly",
		TTLHoursOfMinMaxReplicasRecommendation:     24 * 30,
		MaxReplicasFactor:                          2.0,
		MinReplicasFactor:                          0.5,
		ReplicaReductionFactor:                     0.95,
		UpperTargetResourceUtilization:             90,
		MinimumMinReplicas:                         3,
		PreferredReplicaNumUpperLimit:              30,
		MaximumCPUCores:                            "10",
		MaximumMemoryBytes:                         "10Gi",
		TimeZone:                                   "Asia/Tokyo",
		TortoiseUpdateInterval:                     15 * time.Second,
		TortoiseHPATargetUtilizationMaxIncrease:    5,
		TortoiseHPATargetUtilizationUpdateInterval: time.Hour,
		IstioSidecarProxyDefaultCPU:                "100m",
		IstioSidecarProxyDefaultMemory:             "200Mi",
	}
	if path == "" {
		return config, nil
	}

	// read file from path
	b, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to read config file: %w", err)
	}

	if err := yaml.Unmarshal(b, config); err != nil {
		return nil, fmt.Errorf("failed to unmarshal config file: %w", err)
	}

	if err := validate(config); err != nil {
		return nil, fmt.Errorf("invalid config: %w", err)
	}

	return config, nil
}

func validate(config *Config) error {
	if config.RangeOfMinMaxReplicasRecommendationHours > 24 || config.RangeOfMinMaxReplicasRecommendationHours < 1 {
		return fmt.Errorf("RangeOfMinMaxReplicasRecommendationHours should be between 1 and 24")
	}

	if config.GatheringDataPeriodType != "daily" && config.GatheringDataPeriodType != "weekly" {
		return fmt.Errorf("GatheringDataPeriodType should be either \"daily\" or \"weekly\"")
	}

	if config.TortoiseHPATargetUtilizationMaxIncrease > 100 || config.TortoiseHPATargetUtilizationMaxIncrease <= 0 {
		return fmt.Errorf("TortoiseHPATargetUtilizationMaxIncrease should be between 1 and 100")
	}

	return nil
}
